### PR TITLE
Creates a package publishing script to publish packages as part of the dev roll process.

### PR DESCRIPTION
**Pull request by gspencergoog**
This script will update release metadata in the cloud, and copy the already-built package to the right location and name on cloud storage.

The release metadata will be located in `gs://flutter_infra/releases/releases.json`, and the published packages will end up in `gs://flutter_infra/releases/<channel>/<platform>/flutter_<platform>_<version><archive suffix>`, where `<channel>`, `<platform>`, `<version>`, and `<archive suffix>` are determined by the script.

At the moment, it only supports dev rolls, but (once we know how those will work) should easily support beta rolls as well.

### CODE REVIEW

File: dev/tools/lib/archive_publisher.dart
Comment (goderbauer) [inline dev/tools/lib/archive_publisher.dart]: should this be an enum if it only supports a limited number of pre-defined values?

File: dev/tools/lib/archive_publisher.dart
Comment (goderbauer) [inline dev/tools/lib/archive_publisher.dart]: looks like this throws instead of returning false as documented?

File: dev/tools/lib/archive_publisher.dart
Comment (goderbauer) [inline dev/tools/lib/archive_publisher.dart]: Can we assert on that, e.g. that the hash has at least x digits?

File: dev/tools/test/fake_process_manager.dart
Comment (goderbauer) [inline dev/tools/test/fake_process_manager.dart]: nit: add a doc comment?

File: dev/tools/test/fake_process_manager.dart
Comment (goderbauer) [inline dev/tools/test/fake_process_manager.dart]: same.

**Review by goderbauer — APPROVED** [review]
LGTM modulo some nits

File: dev/tools/lib/roll_dev.dart
Comment (Hixie) [inline dev/tools/lib/roll_dev.dart]: please don't change copyright dates

**Comment by Hixie** [comment]
We rolled v0.0.21 today without this; you may wish to manually update the archive for that.

**Comment by Hixie** [comment]
Is there documentation about how this process works on the wiki? If not, we should add one. In either case, a link to that wiki page would be ideal.

File: dev/tools/lib/archive_publisher.dart
Comment (gspencergoog) [reply dev/tools/lib/archive_publisher.dart]: Done.

File: dev/tools/lib/archive_publisher.dart
Comment (gspencergoog) [reply dev/tools/lib/archive_publisher.dart]: Fixed.

File: dev/tools/lib/archive_publisher.dart
Comment (gspencergoog) [reply dev/tools/lib/archive_publisher.dart]: Yes. Done.

File: dev/tools/test/fake_process_manager.dart
Comment (gspencergoog) [reply dev/tools/test/fake_process_manager.dart]: Good point. I added doc comments for the entire file.

File: dev/tools/test/fake_process_manager.dart
Comment (gspencergoog) [reply dev/tools/test/fake_process_manager.dart]: Done.

File: dev/tools/lib/roll_dev.dart
Comment (Hixie) [reply dev/tools/lib/roll_dev.dart]: It wasn't incorrect. This script is a derived work of a work created in 2017.

Comment (gspencergoog) [reply dev/tools/lib/roll_dev.dart]: Ahh.  Ok, then.  I'll revert it.

**Comment by gspencergoog** [comment]
I used this to publish 0.0.21 (although I called it from a local script instead of roll_dev.dart), so it's available now.  In the future, roll_dev.dart will do it as part of the process.

**Comment by gspencergoog** [comment]
I'll add a section to the release wiki that talks about the publishing process.

**Comment by Hixie** [comment]
Thanks a ton for doing this. I'll wait until this has landed before starting the next roll (tomorrow).

**Pull request merged**
